### PR TITLE
[Bug-fix] Support use of `control` and `adjoint` variants of a kernel

### PR DESCRIPTION
**Pull request by khalatepradnya**
* Handle case where both `control` and `adjoint` variants of a kernel are used.

* Fixes one of the issues captured in issue #1871

### CODE REVIEW

File: test/Quake/control_and_adjoint.qke [hunk 1]
```diff
@@ -0,0 +1,88 @@
+// ========================================================================== //
```
Comment (schweitzpgi) [inline test/Quake/control_and_adjoint.qke:1]: We could merge this with adjoint-1.qke as they are testing the same passes.

File: lib/Optimizer/Transforms/ApplyOpSpecialization.cpp [hunk 2]
```diff
@@ -68,8 +68,14 @@ struct ApplyOpAnalysis {
...
       auto callee = lookupCallee(apply);
       auto iter = infoMap.find(callee);
-      if (iter != infoMap.end())
+      if (iter != infoMap.end()) {
         variant = iter->second;
+        /// NOTE: As per the `DenseMap` API  in
+        /// llvm/include/llvm/ADT/DenseMap.h, an `insert` operation doesn't
+        /// update the value if key is already in the map. Hence, remove it and
+        /// insert a new entry with updated key.
+        infoMap.erase(iter);
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/ApplyOpSpecialization.cpp:77]: I don't think we need to erase anything in the infoMap. Just change the insert line to use `operator[]` and update the map in place. That'll avoid freeing and allocating these structs in the map and be slightly more efficient.

Comment (khalatepradnya) [reply lib/Optimizer/Transforms/ApplyOpSpecialization.cpp]: Addressed in 9e6fc7e3cf41b4d1cd529384078170664446faa7

File: test/Quake/control_and_adjoint.qke (hunk 1 above)
Comment (khalatepradnya) [reply test/Quake/control_and_adjoint.qke:1]: There's another problem in issue#1871 which I was planning to expand in this file as combinations of `control` and `adjoint`, hence prefer to keep it separate.

**Review by schweitzpgi — APPROVED** [review]
+1

**Pull request merged**